### DISCUSSION TITLE
Why do successive requests get read/request timeout?

### DISCUSSION THREAD

**Discussion by forthrin** [HoneyryderChuck/httpx#21] [Q&A]
```ruby
require 'httpx'

timeout = {
  connect_timeout: 99,
  settings_timeout: 99,
  operation_timeout: 99,
  keep_alive_timeout: 99,
  read_timeout: 2,
  write_timeout: 99,
  request_timeout: 3
}
puts HTTPX.get('https://www.example.com/', timeout:).body.to_s.length # OK
sleep(4)
puts HTTPX.get('https://www.example.org/', timeout:).body.to_s.length # ERROR
```

`@error=#<HTTPX::ReadTimeoutError: Timed out after 2 seconds>`

**Comment by HoneyryderChuck** [comment]
hey @forthrin that's definitely a bug. Will have a look.

**Comment by HoneyryderChuck** [comment]
FYI https://gitlab.com/os85/httpx/-/merge_requests/290

**Comment by forthrin** [comment]
Thanks for addressing this quickly. The gem is updated to 1.1.1, and the issue seems to be gone, at first glance at least.

PS! IPv6 is disabled in my macOS network settings. Why do simultaneous IPv6 requests still kick in? Or this is unrelated?

**Reply by HoneyryderChuck** [reply]
I guess you're inspecting the debug logs using level 2? This may mean that you have an ipv6-enabled interface considered valid for "is this machine ipv6-enabled" identification purposes. FWIW this is done [here](https://gitlab.com/os85/httpx/-/blob/master/lib/httpx/options.rb#L19-20), so you may use this snippet to inspect why ruby is considering your system ipv6-enabled.

**Comment by forthrin** [comment]
Since I got the bug, which you write that it was related to simultaneous v4/v6 connection attempts, I assumed the library must have tried to do both v4 and v6. However, the test returns false, so why did I get the bug if only v4 was run on my system?

```ruby
require "socket"
pp(list = Socket.ip_address_list)
pp(list.any? { |a| a.ipv6? && !a.ipv6_loopback? && !a.ipv6_linklocal? && !a.ipv6_unique_local? })
 
[#<Addrinfo: 127.0.0.1>,
 #<Addrinfo: ::1>,
 #<Addrinfo: fe80::1%lo0>,
 #<Addrinfo: fe80::****:****:****:****%ap1>,
 #<Addrinfo: 192.168.1.***>,
 #<Addrinfo: fe80::****:****:****:****%awdl0>,
 #<Addrinfo: fe80::****:****:****:****%llw0>,
 #<Addrinfo: fe80::****:****:****:****%utun0>,
 #<Addrinfo: fe80::****:****:****:****%utun1>,
 #<Addrinfo: fe80::****:****:****:****%utun2>,
 #<Addrinfo: fe80::****:****:****:****%utun3>,
 #<Addrinfo: fe80::****:****:****:****%utun4>]
false
```

This corresponds with the output of `ifconfig`.

The following does not seem to change the result:

- Configure IPv6: Off in System Settings
- Comment out IPv6 entries from `/etc/hosts` and flush DNS

This is not a practical problem in itself, so asking just out of curiosity.

**Comment by HoneyryderChuck** [comment]
The bug was not related with Happy Eyeballs. The description contains 2 separate commit messages, you might be reading the unrelated one.

**Comment by forthrin** [comment]
Ah... Sorted, then.
